File: mcp-server/internal/domains/workflows/fetch.go
package workflows

import (
	"context"
	"fmt"

	workflowv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflow/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	"github.com/stigmer/stigmer/mcp-server/internal/domains"
	"google.golang.org/grpc"
)

// Fetch retrieves a workflow by org and slug, returning its JSON
// representation. Both tool and resource handlers delegate to this function
// for the actual RPC.
func Fetch(ctx context.Context, serverAddress, org, slug string) (string, error) {
	return domains.WithConnection(ctx, serverAddress,
		func(ctx context.Context, conn *grpc.ClientConn) (string, error) {
			client := workflowv1.NewWorkflowQueryControllerClient(conn)
			workflow, err := client.GetByReference(ctx, &apiresource.ApiResourceReference{
				Org:  org,
				Kind: apiresourcekind.ApiResourceKind_workflow,
				Slug: slug,
			})
			if err != nil {
				return "", domains.RPCError(err, fmt.Sprintf("workflow %q in org %q", slug, org))
			}
			return domains.MarshalJSON(workflow)
		})
}

File: mcp-server/src/domains/workflows/fetch.ts
// Workflow read path: the single RPC both the get_workflow tool and the
// workflow resource template delegate to.
// Go parity: mcp-server/internal/domains/workflows/fetch.go.

import { WorkflowSchema } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/api_pb";
import { WorkflowQueryController } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/query_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import { withClient } from "../client.js";
import { toProtoJson } from "../marshal.js";
import { rpcError } from "../rpcerr.js";

/**
 * Retrieve a workflow by org and slug, returning its protojson representation.
 * Errors are classified into user-facing messages via {@link rpcError}.
 */
export async function fetchWorkflow(
  serverAddress: string,
  token: string,
  org: string,
  slug: string,
): Promise<string> {
  return withClient(WorkflowQueryController, serverAddress, token, async (client, callOptions) => {
    try {
      const workflow = await client.getByReference(
        { org, kind: ApiResourceKind.workflow, slug },
        callOptions,
      );
      return toProtoJson(WorkflowSchema, workflow);
    } catch (err) {
      throw rpcError(err, `workflow "${slug}" in org "${org}"`);
    }
  });
}
